ONSdigital/rm-collection-exercise-service
# Pass sampleUnitId through to case service

# Motivation and Context
Originally this was going to be a simple change to include the sample UUID (newly passed in the message from the sample service) in the message that gets passed in the case service.  

However, we strongly felt that an integration test should be created to test these changes (i.e. one where the sample message is passed in by the test which then asserts that a message has been sent to the case service).

In order for the message to be sent to the case service, the sample must be fully validated.  Also the collection exercise must be in the VALIDATED state, which requires it to be executed first.  For this to work end to end, the collection exercise service needed to be changed to fully support social surveys.  In the end, this required surprisingly little code but much investigation (details in the next section).

# What has changed
- sample unit id (UUID) now received in message from sample service and persisted in the collection exercise service internal representation
- when a collection exercise is created for a social survey, the correct action plans, case type defaults etc are now created
- for social surveys, no attempt is made to get a party id for the incoming samples 
- samples for social surveys are validated by mandating a collection instrument but insisting that party is absent (null)
- a new integration test ([ensureSampleUnitIdIsPropagatedHere](https://github.com/ONSdigital/rm-collection-exercise-service/blob/5e2c233d5ee7102aa56fcafa18ef9ba4be2d547b/src/test/java/uk/gov/ons/ctp/response/collection/exercise/endpoint/CollectionExerciseEndpointIT.java#L207)) has been added to test these changes end to end 
- mocking using wiremock has been added to allow the collection exercise service to get details of collection instrument (party service response also added to get tests to pass initially but not currently used - this has been left in because we will need to make a similar IT for business surveys and it will be needed for that)
- test data and unit tests have been updated to account for the additional sample unit id

# How to test?
- check out the code
- run 'mvn clean install'
- ensure all tests pass

# TODO 
- implement similar integration test for business surveys

# Links
[New integration test](https://github.com/ONSdigital/rm-collection-exercise-service/blob/5e2c233d5ee7102aa56fcafa18ef9ba4be2d547b/src/test/java/uk/gov/ons/ctp/response/collection/exercise/endpoint/CollectionExerciseEndpointIT.java#L207)
[WireMock](http://wiremock.org/)
[Trello card](https://trello.com/c/ZFF7x8Za/102-dev-task-3b-int-create-action-service-to-retrieve-new-sample-attributes-13)
